### ISSUE TITLE
Query with array join and toYYYYMM in predicate doesn't work as expected

### ISSUE THREAD

**Issue by canhld94** [ClickHouse/ClickHouse#53152]
Reproduce:

```sql
select
  toYYYYMM(date) as date_,
  n
from (select 
  [toDate(now()), toDate(now())] as date,
  [1, 2] as n
) as data
array join date, n
where date_ >= 202303;
```

Work in 23.6: https://fiddle.clickhouse.com/cabc075a-ac8a-4a15-bd46-2bc2b1de09a6
Not work in 23.7: https://fiddle.clickhouse.com/74c0b0b6-72f2-4fb7-8a0b-bc098307da70

Suspicious PR: #52091

**Comment by ZhiguoZh** [comment]
Hi @canhld94, thanks for reporting this issue! I'll start working on its fix right away!

**Comment by ZhiguoZh** [comment]
Hi @canhld94, the reported issue has been fixed in #53440, and the problematic query has been added in the CI test. Thanks!
